compiler/rustc_parse/src/parser/stmt.rs

```diff
@@ -38,30 +40,47 @@ impl<'a> Parser<'a> {
         capture_semi: bool,
         force_collect: ForceCollect,
     ) -> PResult<'a, Option<Stmt>> {
-        let mut attrs = self.parse_outer_attributes()?;
+        let attrs = self.parse_outer_attributes()?;
         let lo = self.token.span;
 
-        maybe_whole!(self, NtStmt, |stmt| {
-            let mut stmt = stmt;
-            stmt.visit_attrs(|stmt_attrs| {
-                mem::swap(stmt_attrs, &mut attrs);
-                stmt_attrs.extend(attrs);
-            });
-            Some(stmt)
-        });
+        // Don't use `maybe_whole` so that we have precise control
+        // over when we bump the parser
+        if let token::Interpolated(nt) = &self.token.kind {
+            if let token::NtStmt(stmt) = &**nt {
+                let mut stmt = stmt.clone();
+                return self.collect_tokens_trailing_token(
+                    attrs,
+                    force_collect,
+                    |this, mut attrs| {
+                        stmt.visit_attrs(|stmt_attrs| {
+                            mem::swap(stmt_attrs, &mut attrs);
+                            stmt_attrs.extend(attrs);
+                        });
+                        // Make sure we capture the token::Interpolated
```

Comment:
Is this necessary?
The interpolated token is going to always be replaced by its underlying tokens.

Reply:
Same applies to `parse_item_common`.

Reply:
Currently, I don't think this is strictly necessary, since we'll end up with the same attributes and tokens even if we don't call `collect_tokens_trailing_token`.

However, this will become necessary when we add `PreexpTokenStream` - calling `collect_tokens_trailing_token` will ensure that we properly insert any new attributes into the `PreexpTokenStream`. That is, we need `#[cfg(FALSE)] $item` to get removed from the `PreexpTokenStream`.

Alternatively, we could remove this logic, and directly manipulate the `PreexpTokenStream` (once it gets added). However, I think it's cleaner to have all of the token-related attribute handling go through `collect_tokens_trailing_token`.